libparsec/crates/platform_storage/src/web/user.rs

```diff
@@ -2,76 +2,122 @@
 
 use std::{path::Path, sync::Arc};
 
+use indexed_db_futures::prelude::*;
 use libparsec_types::prelude::*;
 
+use crate::web::model::{RealmCheckpoint, Vlob};
+
 #[allow(unused)]
 pub struct NeedSyncEntries {
     pub remote: Vec<VlobID>,
     pub local: Vec<VlobID>,
 }
 
-#[allow(unused)]
 #[derive(Debug)]
-pub struct UserStorageUpdater {}
-
-#[allow(unused)]
-impl UserStorageUpdater {
-    pub async fn set_user_manifest(
-        &self,
-        _user_manifest: Arc<LocalUserManifest>,
-    ) -> anyhow::Result<()> {
-        todo!();
-    }
+pub struct PlatformUserStorage {
+    conn: Arc<IdbDatabase>,
+    realm_id: VlobID,
 }
 
-#[derive(Debug)]
-pub struct PlatformUserStorage {}
+// Safety: PlatformUserStorage is read only
+unsafe impl Send for PlatformUserStorage {}
 
 impl PlatformUserStorage {
-    #[allow(dead_code)]
-    pub async fn start(_data_base_dir: &Path, _device: Arc<LocalDevice>) -> anyhow::Result<Self> {
-        todo!();
-    }
-
     pub(crate) async fn no_populate_start(
         _data_base_dir: &Path,
-        _device: &LocalDevice,
+        device: &LocalDevice,
     ) -> anyhow::Result<Self> {
-        todo!();
+        // 1) Open the database
+
+        let db_req =
+            IdbDatabase::open_u32("database_name", 1).map_err(|e| anyhow::anyhow!("{e:?}"))?;
```

Comment:
You likely need to setup an handle on for the event [`IdbVersionChangeEvent`](https://docs.rs/indexed_db_futures/latest/indexed_db_futures/prelude/struct.IdbVersionChangeEvent.html)

If configured that way, you rewrite `initialize_model_if_needed` to be a migration from `v0 (uninitialized) -> v1 (initialized)`.
That would also simplify the said function as you won't need to check if the model exist.

Reply:
`versionchange` is send when the event `upgradeneeded` is send or when db is deleted.

<https://developer.mozilla.org/en-US/docs/Web/API/IDBDatabase/versionchange_event>

 But having a handle on [`upgradeneeded`](https://developer.mozilla.org/en-US/docs/Web/API/IDBOpenDBRequest/upgradeneeded_event) seems to be what we need